cirq/circuits/frozen_circuit.py

```diff
@@ -85,13 +89,19 @@ def moments(self) -> Sequence['cirq.Moment']:
     def device(self) -> devices.Device:
         return self._device
 
+    @property
+    def name(self) -> Optional[str]:
+        return self._name
+
     def __hash__(self):
-        return hash((self.moments, self.device))
+        return hash((self.moments, self.device, self.name))
+
+    def _serialization_name_(self):
+        return self._name or 'Circuit'
 
-    def serialization_key(self):
-        # TODO: use this key in serialization and support user-specified keys.
+    def diagram_name(self):
         key = hash(self) & 0xFFFF_FFFF_FFFF_FFFF
-        return f'Circuit_0x{key:016x}'
+        return f'{self.name or "Circuit"}_0x{key:016x}'
```

Comment:
Should `self.name or "Circuit"` be `self._serialization_name_()` for consistency, or do we want to keep them independent and they just happen to match today?

Reply:
It's plausible that we'll want to separate these in the future (e.g. for out-of-line subcircuits in circuit diagrams) - let's keep them independent.